src/services/db/GitFileSystemService.ts

```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
     })
   }
 
-  // TODO: Creates either directory or file
-  // ResourceDirectoryService used this to create a directory + file at the same time
-  create() {}
+  // Creates a file and the associated directory if it doesn't exist
+  create({
+    repoName,
+    userId,
+    content,
+    directoryName,
+    fileName,
+  }: {
+    repoName: string
+    userId: string
+    content: string
+    directoryName: string
+    fileName: string
+  }): ResultAsync<
+    { sha: string },
+    ConflictError | GitFileSystemError | NotFoundError
+  > {
+    const filePath = directoryName ? `${directoryName}/${fileName}` : fileName
+    const pathToEfsDir = `${EFS_VOL_PATH}/${repoName}/${directoryName}/`
+    const pathToEfsFile = `${EFS_VOL_PATH}/${repoName}/${filePath}`
+    const encodedContent = content
+    let oldStateSha = ""
+
+    return this.getLatestCommitOfBranch(repoName, BRANCH_REF)
+      .andThen((latestCommit) => {
+        const { sha } = latestCommit
+        if (!sha) {
+          return errAsync(new GitFileSystemError("An unknown error occurred"))
+        }
+        oldStateSha = sha
+        return okAsync(true)
+      })
+      .andThen(() => this.getFilePathStats(repoName, directoryName))
+      .andThen(() => ok(""))
+      .orElse((error) => {
+        if (error instanceof NotFoundError) {
+          // Create directory if it does not already exist
+          try {
+            fs.promises.mkdir(pathToEfsDir)
+            return ok("")
+          } catch (mkdirErr) {
+            logger.error(
+              `Error occurred while creating ${pathToEfsDir} directory: ${mkdirErr}`
+            )
+            return err(new GitFileSystemError("An unknown error occurred"))
+          }
+        }
+        return err(error)
+      })
+      .andThen(() => this.getFilePathStats(repoName, filePath))
+      .andThen((stats) => {
+        if (stats.isFile())
+          return err(
+            new ConflictError(
+              `File ${filePath} already exists in repo ${repoName}`
+            )
+          )
+        return ok("")
+      })
+      .orElse((error) => {
+        if (error instanceof NotFoundError) {
+          return ok("")
+        }
+        return err(error)
+      })
+      .andThen(() =>
+        ResultAsync.fromPromise(
+          fs.promises.writeFile(pathToEfsFile, encodedContent),
+          (error) => {
+            logger.error(`Error when creating ${filePath}: ${error}`)
+            if (error instanceof Error) {
+              return new GitFileSystemError(error.message)
+            }
+            return new GitFileSystemError("An unknown error occurred")
```

Comment:
We potentially might need a rollback here.

Reply:
Fixed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/a1dd5c2b37fc75410e1f58d9c9c9d4a4f2cde9a4